content/blog/elastic-apm-plugin.md

```diff
@@ -0,0 +1,113 @@
+
+We can see which HTTP request are fired and what database calls are being made. This information can be extremely useful when trying to understand what's happening within your application, for example when looking into a performance issue.
+
+### What if your framework is not supported?
+
+But what to do when your framework of choice is not supported by the agent? By not doing anything you might be missing out on valuable information, so is there anything we can do? Luckily there is! Most APM agents will offer some way of adding manual instrumentation in your code. This might be suitable if you want to add some instrumentation for a very specific use case. But what if for example your database framework is not supported? Adding manual instrumentation for each query is not an ideal solution.
+
+Today we'll be looking at an example were a Java application was using R2DBC (a database framework) and Elastic as the APM provider. The Elastic APM agent offers a plugin api, so we can write our own instrumentation without needing to instrument each individual query. 
+
+### How does the APM agent work?
+
+Before we look into how we can write a plugin for our example case, let's dive into how the Elastic APM agent works. The Elastic APM agent uses bytecode manipulation to instrument code, using bytecode manipulation it can modify Java classes at runtime. This way the agent can change a class without recompiling it. What will typically happen is that the agent will add some code when an instrumented method is entered and exited. 
+
+A good example to demonstrate how this works in practice is the Servlet API. The Servlet API in Java is the main entry point for most HTTP servers. There are many implementations, but by instrumentation on the API level it doesn't matter which implementation is used. The main entry point for the Servlet API is the `service` method[^1]. We could write instrumentation which would indicate the start of HTTP request the moment the `service` method is entered and indicate the end of the HTTP request the method is exited.
+
+### Let's write our R2DBC plugin
+
+We now have a basic idea of how the APM agent works, so we can have a look at how we could write a plugin for our problem at hand. The aim is to be able to record queries that have been executed, including SQL statement. We first need to identify what our entry point will be to instrument. The most obvious choice seems to be `io.r2dbc.spi.Statement#execute[^2]. According to the javadoc this method is responsible for `Executes one or more SQL statements and returns the Results.`. There is one problem however, the `Statement` API doesn't have any reference to the SQL statement. There are probably some ways around this, but to keep it simple we'll instrument a specific implementation: `io.r2dbc.postgresql.PostgresqlStatement`. This implementation has a `execute` method which takes the SQL statement as parameter. 
```

Comment:
I believe there's a backtick missing here, right before "According to the javadoc"